plugins/inputs/ping/ping.go

```diff
@@ -194,137 +177,153 @@ func hostPinger(binary string, timeout float64, args ...string) (string, error)
+		timeout = 5
+	}
+
+	tick := time.NewTicker(time.Duration(interval * float64(time.Second)))
+	defer tick.Stop()
+
 	if p.Deadline > 0 {
-		switch system {
-		case "darwin", "freebsd", "netbsd", "openbsd":
-			args = append(args, "-t", strconv.Itoa(p.Deadline))
-		case "linux":
-			args = append(args, "-w", strconv.Itoa(p.Deadline))
-		default:
-			// not sure the best option here, just assume gnu ping?
-			args = append(args, "-w", strconv.Itoa(p.Deadline))
-		}
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(p.Deadline)*time.Second)
+		defer cancel()
 	}
-	if p.Interface != "" {
-		switch system {
-		case "darwin":
-			args = append(args, "-I", p.Interface)
-		case "freebsd", "netbsd", "openbsd":
-			args = append(args, "-S", p.Interface)
-		case "linux":
-			args = append(args, "-I", p.Interface)
-		default:
-			// not sure the best option here, just assume gnu ping?
-			args = append(args, "-i", p.Interface)
-		}
+
+	chanLength := 100
+	if p.Count > 0 {
+		chanLength = p.Count
 	}
-	args = append(args, url)
-	return args
-}
 
-// processPingOutput takes in a string output from the ping command, like:
-//
-//     ping www.google.com (173.194.115.84): 56 data bytes
-//     64 bytes from 173.194.115.84: icmp_seq=0 ttl=54 time=52.172 ms
-//     64 bytes from 173.194.115.84: icmp_seq=1 ttl=54 time=34.843 ms
-//
-//     --- www.google.com ping statistics ---
-//     2 packets transmitted, 2 packets received, 0.0% packet loss
-//     round-trip min/avg/max/stddev = 34.843/43.508/52.172/8.664 ms
-//
-// It returns (<transmitted packets>, <received packets>, <average response>)
-func processPingOutput(out string) (int, int, int, float64, float64, float64, float64, error) {
-	var trans, recv, ttl int = 0, 0, -1
-	var min, avg, max, stddev float64 = -1.0, -1.0, -1.0, -1.0
-	// Set this error to nil if we find a 'transmitted' line
-	err := errors.New("Fatal error processing ping output")
-	lines := strings.Split(out, "\n")
-	for _, line := range lines {
-		// Reading only first TTL, ignoring other TTL messages
-		if ttl == -1 && strings.Contains(line, "ttl=") {
-			ttl, err = getTTL(line)
-		} else if strings.Contains(line, "transmitted") &&
-			strings.Contains(line, "received") {
-			trans, recv, err = getPacketStats(line, trans, recv)
-			if err != nil {
-				return trans, recv, ttl, min, avg, max, stddev, err
-			}
-		} else if strings.Contains(line, "min/avg/max") {
-			min, avg, max, stddev, err = checkRoundTripTimeStats(line, min, avg, max, stddev)
-			if err != nil {
-				return trans, recv, ttl, min, avg, max, stddev, err
-			}
+	resps := make(chan *ping.Response, chanLength)
+	rsps := []*ping.Response{}
+
+	r := &sync.WaitGroup{}
+	r.Add(1)
+	go func() {
+		for res := range resps {
+			rsps = append(rsps, res)
+		}
+		r.Done()
+	}()
+
+	packetsSent := 0
+	wg := &sync.WaitGroup{}
+	c := ping.Client{}
+
+	for p.Count <= 0 || packetsSent < p.Count {
+		select {
+		case <-ctx.Done():
+			goto finish
```

Comment:
Just an observation, you don't need to change anything... To me this is an indication flag that this function is too big.  If you split it up in the right spots I think error handling could be easier, less variable name aliasing, and no more ugly goto.

Reply:
yes.. I'd love any pointers if you find yourself with some time